[GCI] Add test for extract_submission_data

nikochiko opened the pull request:
* As part of a GCI task.
* Added `test_extract_submission_data` to test function [`extract_submission_data` ](https://github.com/Cloud-CV/EvalAI/blob/master/scripts/workers/submission_worker.py#L275) from submission worker.

Ram81 commented on tests/unit/worker/test_submission_worker.py:
```diff
@@ -45,6 +123,18 @@ def test_return_file_url_per_environment(self):
+    def test_extract_submission_data(self, mocked_download_and_extract_zip_file, mocked_import_module):
+        phases = self.challenge.challengephase_set.all()
+        extract_challenge_data(self.challenge, phases)
+        evaluation_script_url = "http://testserver{}".format(self.challenge.evaluation_script.url)
```
can we create a common variable for `http://testserver`, hardcoding it at multiple places in a file is a bad practice.

nikochiko replied:
Okay right. Made changes, please see new commit.

Ram81 replied:
@nikochiko can you please fix the reviews in your previous PR? we need to merge it before getting to this PR as it has redundant code

nikochiko replied:
That pull is almost done as well. Made some changes here to update it with base changes in previous PR like changing `unittest.TestCase` with `rest_framework.APITestCase`.

Ram81 commented on tests/unit/worker/test_submission_worker.py:
```diff
@@ -30,6 +45,71 @@ def setUp(self):
+            published=False,
+            enable_forum=True,
+            anonymous_leaderboard=False,
+            max_concurrent_submission_evaluation=200000,
```
@nikochiko we need to rebase the changes which are already there in your other PR. For now can you make the same changes here. For example changing `max_concurrent_submission_evaluation=100`

Ram81 commented on tests/unit/worker/test_submission_worker.py:
```diff
@@ -30,6 +45,71 @@ def setUp(self):
+            ),
+            max_submissions_per_day=100000,
+            max_submissions_per_month=100000,
+            max_submissions=100000,
```
@nikochiko can you reduce these numbers too (smaller range)

nikochiko replied:
I just copy-pasted this from another file. Will make the changes.

Ram81 commented on tests/unit/worker/test_submission_worker.py:
```diff
@@ -30,6 +45,71 @@ def setUp(self):
+            ),
+            method_name="Test Method 1",
+            method_description="Test Description 1",
+            project_url="http://testserver1/",
```
Why can't we use `self.testserver` here?

nikochiko replied:
Same, I copy-pasted this. Will make the edit.

Ram81 commented on tests/unit/worker/test_submission_worker.py:
```diff
@@ -43,7 +123,27 @@ def test_create_dir_as_python_package(self):
+    @mock.patch("scripts.workers.submission_worker.SUBMISSION_INPUT_FILE_PATH", "/tmp/test-dir/compute/submission_files/submission_{submission_id}/{input_file}")
+    @mock.patch("scripts.workers.submission_worker.create_dir_as_python_package")
+    @mock.patch("scripts.workers.submission_worker.download_and_extract_file")
+    def test_extract_submission_data_successful(self, mock_download_and_extract_file, mock_create_dir_as_python_package):
```
can you rename the method to `test_extract_submission_data` or `test_extract_submission_data_success`

Ram81 commented on tests/unit/worker/test_submission_worker.py:
```diff
@@ -43,7 +123,27 @@ def test_create_dir_as_python_package(self):
+    @mock.patch("scripts.workers.submission_worker.download_and_extract_file")
+    def test_extract_submission_data_successful(self, mock_download_and_extract_file, mock_create_dir_as_python_package):
+        submission = extract_submission_data(self.submission.pk)
+        expected_submission_data_dir = "/tmp/test-dir/compute/submission_files/submission_{submission_id}".format(submission_id=self.submission.pk)
```
submission path prefix is being used at multiple places can we create a class variable for this `tmp/test-dir/compute/submission_files`

nikochiko replied:
Then is it okay to add class variables for all paths [as in `submission_worker.py`](https://github.com/Cloud-CV/EvalAI/blob/master/scripts/workers/submission_worker.py#L56-L66)? It will help for creating other tests like that for `extract_challenge_data`.

nikochiko commented:
@Ram81 please review the changes.
* I have added more class variables replicating those from [`submission_worker.py`](https://github.com/Cloud-CV/EvalAI/blob/master/scripts/workers/submission_worker.py#L56-L66). Is it okay that I have replicated all class variables, even those not used in the tests? They can be used for writing missing tests.
* Changed `mock.patch` structure to accommodate this change. It could not be used in decorator.

Ram81 commented on tests/unit/worker/test_submission_worker.py:
```diff
@@ -1,28 +1,77 @@
+        self.CHALLENGE_DATA_BASE_DIR = join(
+            self.COMPUTE_DIRECTORY_PATH, "challenge_data"
+        )
+        self.SUBMISSION_DATA_BASE_DIR = join(
```
@nikochiko I think creating just 1 variable for `/compute/challenge_data/submission_files` prefix is enough. I don't think we will need separate variables for `CHALLENGE_DATA_BASE_DIR` or `COMPUTE_DIRECTORY_PATH` for now.

Ram81 commented on tests/unit/worker/test_submission_worker.py:
```diff
@@ -1,28 +1,77 @@
+        self.PHASE_DATA_BASE_DIR = join(
+            self.CHALLENGE_DATA_DIR, "phase_data"
+        )
+        self.PHASE_DATA_DIR = join(
```
@nikochiko rather than coding in this way you can add helper methods which will take `phase_id` as input and return the complete path. I don't think we will need `PHASE_DATA_BASE_DIR` var separately. You can write methods to get path based on `phase_id` and `submission_id`

Ram81 commented on tests/unit/worker/test_submission_worker.py:
```diff
@@ -43,7 +162,36 @@ def test_create_dir_as_python_package(self):
+        mock_create_dir_as_python_package.assert_called_with(expected_submission_data_dir)
+
+        expected_submission_input_file = "{0}{1}".format(self.testserver, self.submission.input_file.url)
+        expected_submission_input_file_path = self.SUBMISSION_INPUT_FILE_PATH.format(submission_id=self.submission.pk, input_file=os.path.basename(self.submission.input_file.name))
```
You can use the method to fetch `SUBMISSION_INPUT_FILE_PATH` for a specific submission id. This would be more modular

Ram81 commented:
@nikochiko your 1st PR is merged into our master branch which is why you are seeing merge conflicts with this PR. Please rebase your PR with the master branch and resolve the merge conflicts. If you need any help from my end to resolve conflicts I'll be happy to help 😄

nikochiko commented:
@Ram81 , thanks for your insightful reviews. I am already learning a lot. Glad to know you offered to help, I have some doubts and will ping you on Gitter about that. :D

Ram81 commented:
Sure

nikochiko commented:
I added the requested changes and a few more: 
* Add helper methods `helper_return_submission_input_file_path` and `helper_return_phase_annotation_file_path`. 
* Renamed `patcher_submision_input_file_path` and `patcher_submission_data_dir` to `mock_submission_input_file_path` and `mock_submission_data_dir`. 
* Remove extra variables. 
@Ram81 Any improvements that can be made? Please review.

Ram81 commented on tests/unit/worker/test_submission_worker.py:
```diff
@@ -67,6 +90,72 @@ def setUp(self):
+            is_flagged=True,
+        )
+
+    def helper_return_submission_input_file_path(self, submission_id, input_file):
```
@nikochiko can we rename the method to `get_submission_input_file_path`

nikochiko replied:
Yeah, that would be more friendly. Will add the change.

Ram81 commented on tests/unit/worker/test_submission_worker.py:
```diff
@@ -67,6 +90,72 @@ def setUp(self):
+            input_file=input_file_name,
+        )
+
+    def helper_return_phase_annotation_file_path(
```
Same here can we rename this to `get_phase_annotation_file_path`

Ram81 commented on tests/unit/worker/test_submission_worker.py:
```diff
@@ -79,7 +168,51 @@ def test_create_dir_as_python_package(self):
+            self,
+            mock_download_and_extract_file,
+            mock_create_dir_as_python_package):
+        submission_input_file_path = join(
```
As I can see that we only use `SUBMISSION_DATA_DIR` with `{input_file}` so can we define this directly in it's declaration. As this line is redundant to line 139

nikochiko replied:
We need it to mock `scripts.worker.submission_worker.SUBMISSION_DATA_DIR`, I think you missed that it has been used on [line 182](https://github.com/Cloud-CV/EvalAI/pull/2498/files#diff-1aeba0ea1ecc134c604f070c6be1385aR182). I think we should keep it as it is.

Ram81 commented on tests/unit/worker/test_submission_worker.py:
```diff
@@ -67,6 +90,72 @@ def setUp(self):
+            challenge_id=challenge_id,
+            phase_id=phase_id,
+            annotation_file=annotation_file_name,
+        )
```
Why is this method not being used anywhere?

nikochiko replied:
Yes. While I was importing the class variables from `submission_worker.py`, I thought it would be good to import all of them since there are other missing tests as well. The other tests such as that for `extract_challenge_data` would use these additions. It is in the GCI tasks so it would be helpful then. Also, it had almost the same template as the other helper method, so I thought I could add it now itself.

nikochiko replied:
I have made these gists: https://gist.github.com/nikochiko/c259554d3b2d2ea5cbea3fdfa65576f1 and https://gist.github.com/nikochiko/de6265470affbe221b41e3f77d46c09c for future reference. Will remove the unused methods and class variables for now.

nikochiko commented:
@Ram81 I had placed some unused variables and that method for future use, because they were similar to the changes being made in this PR. But, I think it is best that they be added later so we can avoid having unused code. I removed them and also renamed the helper method as requested. Please review.
